src/game/logic/object/weapon.cpp

```diff
@@ -954,6 +1520,214 @@ void WeaponTemplate::Deal_Damage_Internal(ObjectID source_id,
+                        if ((mask & AFFECTS_SELF) == 0) {
+                            if (source == obj) {
+                                goto next;
+                            }
+
+                            if (source->Get_Producer_ID() == obj->Get_ID()) {
+                                goto next;
+                            }
+                        }
+
+                        if ((mask & AFFECTS_NOT_SIMILAR) != 0) {
+                            if (source->Get_Template()->Is_Equivalent_To(obj->Get_Template())) {
+                                if (source->Get_Relationship(obj) == ALLIES) {
+                                    goto next;
+                                }
+                            }
+                        }
+
+                        if ((mask & AFFECTS_NOT_AIRBORNE) != 0 && obj->Is_Significantly_Above_Terrain()) {
+                            goto next;
+                        }
+
+                        Relationship relationship = source->Get_Relationship(obj);
+                        WeaponAffectsType affects;
+
+                        if (relationship == ALLIES) {
+                            affects = AFFECTS_ALLIES;
+                        } else if (relationship == ENEMIES) {
+                            affects = AFFECTS_ENEMIES;
+                        } else {
+                            affects = AFFECTS_NEUTRALS;
+                        }
+
+                        if ((affects & mask) == 0) {
+                            goto next;
+                        }
+                    }
+                }
+
+                info.m_in.m_damageType = damage;
+                info.m_in.m_deathType = death;
+                info.m_in.m_sourceID = source_id;
+                info.m_in.m_playerMask = 0;
+                info.m_in.m_damageStatusType = status;
+                pos2.Zero();
+
+                if (obj != nullptr && source != nullptr) {
+                    pos2.Set(obj->Get_Position());
+                    pos2.Sub(source->Get_Position());
+                }
+
+                angle = Get_Radius_Damage_Angle();
+
+                if ((angle >= DEG_TO_RADF(180.0f) || obj != nullptr) && source != nullptr) {
+                    Vector3 v1;
+                    source->Get_Transform_Matrix()->Get_X_Vector(&v1);
+                    Vector3 v2(pos2.x, pos2.y, pos2.z);
+                    v1.Normalize();
+                    v2.Normalize();
+                    if (GameMath::Cos(angle) <= v1 * v2) {
+                        info.m_in.m_shockWaveAmount = m_shockWaveAmount;
+
+                        if (info.m_in.m_shockWaveAmount > 0.0f) {
+                            if (GameMath::Fabs(pos2.x) < GAMEMATH_EPSILON && GameMath::Fabs(pos2.y) < GAMEMATH_EPSILON
+                                && GameMath::Fabs(pos2.z) < GAMEMATH_EPSILON) {
+                                pos2.z = 1.0f;
+                            }
+
+                            info.m_in.m_shockWavePos = pos2;
```

Comment:
This looks odd. pos2 is a direction vector (aka local position), not a world position. See above
```
                    pos2.Set(obj->Get_Position());
                    pos2.Sub(source->Get_Position());
```
Are you sure this is correct? If it is correct, then I assume `m_shockWavePos` is a relative position.

Reply:
This is correct.